narwhal/worker/src/client.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    net::Ipv4Addr,
+    sync::{Arc, Mutex},
+};
+
+use arc_swap::ArcSwap;
+use mysten_network::{multiaddr::Protocol, Multiaddr};
+use thiserror::Error;
+use types::{metered_channel::Sender, Transaction, TxResponse};
+
+/// Uses a map to allow running multiple Narwhal instances in the same process.
+/// TODO: after Rust 1.66, use BTreeMap::new() instead of wrapping it in an Option.
+static LOCAL_NARWHAL_CLIENTS: Mutex<Option<BTreeMap<Multiaddr, Arc<ArcSwap<LocalNarwhalClient>>>>> =
+    Mutex::new(None);
+
+/// The maximum allowed size of transactions into Narwhal.
+/// TODO: maybe move to TxValidator?
+pub const MAX_ALLOWED_TRANSACTION_SIZE: usize = 6 * 1024 * 1024;
+
+/// Errors returned to clients submitting transactions to Narwhal.
+#[derive(Clone, Debug, Error)]
+pub enum NarwhalError {
+    #[error("Failed to include transaction in a header!")]
+    TransactionNotIncludedInHeader,
+
+    #[error("Narwhal is shutting down!")]
+    ShuttingDown,
+
+    #[error("Transaction is too large: size={0} limit={1}")]
+    TransactionTooLarge(usize, usize),
+}
+
+/// TODO: add NarwhalClient trait and implement RemoteNarwhalClient with grpc.
+
+/// A client that connects to Narwhal locally.
+#[derive(Clone)]
+pub struct LocalNarwhalClient {
+    /// TODO: maybe use tx_batch_maker for load schedding.
+    tx_batch_maker: Sender<(Transaction, TxResponse)>,
+}
+
+impl LocalNarwhalClient {
+    pub fn new(tx_batch_maker: Sender<(Transaction, TxResponse)>) -> Arc<Self> {
+        Arc::new(Self { tx_batch_maker })
+    }
+
+    /// Sets the instance of LocalNarwhalClient for the local address.
+    /// Address is only used as the key.
+    pub fn set_global(addr: Multiaddr, instance: Arc<Self>) {
+        let addr = Self::canonicalize_address_key(addr);
```

Comment:
As I said above we could probably avoid dealing with addresses and just used instead the `worker_id`. I understand that it might be an issue for any current tests that use multiple authorities thus worker_id would create some override - so probably you used the addr instead? Maybe if we end up with the plumbing solution we won't have to deal with those details anyway.